### PR TITLE
IT test for resolve-test-dependencies with snapshots

### PR DESCRIPTION
**Pull request by jtnord**
this introduces test coverage for both `resolve-test-dependencies` with `overrideWar` containing a `SNAPSHOT` plugin as well as custom-war with `addThisPluginToCustomWar`

writing the IT was not done separately one for each PR or moved as this would have created more renamed files bloating the diff, or duplicating tests unnecessarily.  The commits here are separate for each fix - but review in the respective PRs and the checks and order demonstrates each fix in turn)

* the IT here is the first commit  694b0cdcf533e320b0befe92e136a2db02d985ce
* the second e493f342e0da44570ecee2d9a77e5a4c86bc96a6  is #481
* the third commit 210884d2a6f1aedd6df012ec1e525759529917bc is #478 

### Testing done

* [failure](https://github.com/jenkinsci/maven-hpi-plugin/pull/480/checks?check_run_id=13268354582) without any fixes (#479)
```
io.jenkins.tools.maven_hpi_plugin.its.override-war-with-snapshot.override-war-with-snapshot
Assertion failed: 

assert jf.getEntry('WEB-INF/plugins/override-war-with-snapshot.hpi') != null
       |  |                                                          |
       |  null                                                       false
       java.util.jar.JarFile@78a96c28
```

* [failure](https://github.com/jenkinsci/maven-hpi-plugin/pull/480/checks?check_run_id=13269115795) with the custom-war fix (#481)
```
[INFO] --- hpi:3.44-SNAPSHOT:resolve-test-dependencies (default-cli) @ dependant ---
[INFO] Scanned contents of /home/jenkins/agent/workspace/Plugins_maven-hpi-plugin_PR-480/target/its/override-war-with-snapshot/war-with-plugin/target/override-war-with-snapshot.war with 1 bundled plugins
[INFO] Updating direct dependency org.jenkins-ci.tools.hpi.its:override-war-with-snapshot from 1.0.0-SNAPSHOT to 2.0.0-SNAPSHOT (private-e493f342-jenkins)
[INFO] Downloading from do-proxy-incrementals: https://repo.do.jenkins.io/incrementals/org/jenkins-ci/tools/hpi/its/override-war-with-snapshot/2.0.0-SNAPSHOT%20(private-e493f342-jenkins)/override-war-with-snapshot-2.0.0-SNAPSHOT%20(private-e493f342-jenkins).pom
[INFO] Downloading from do-proxy: https://repo.do.jenkins.io/public/org/jenkins-ci/tools/hpi/its/override-war-with-snapshot/2.0.0-SNAPSHOT%20(private-e493f342-jenkins)/override-war-with-snapshot-2.0.0-SNAPSHOT%20(private-e493f342-jenkins).pom
[WARNING] The POM for org.jenkins-ci.tools.hpi.its:override-war-with-snapshot:jar:2.0.0-SNAPSHOT (private-e493f342-jenkins) is missing, no dependency information available
[INFO] Downloading from do-proxy-incrementals: https://repo.do.jenkins.io/incrementals/org/jenkins-ci/tools/hpi/its/override-war-with-snapshot/2.0.0-SNAPSHOT%20(private-e493f342-jenkins)/override-war-with-snapshot-2.0.0-SNAPSHOT%20(private-e493f342-jenkins).jar
[INFO] Downloading from do-proxy: https://repo.do.jenkins.io/public/org/jenkins-ci/tools/hpi/its/override-war-with-snapshot/2.0.0-SNAPSHOT%20(private-e493f342-jenkins)/override-war-with-snapshot-2.0.0-SNAPSHOT%20(private-e493f342-jenkins).jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.668 s
[INFO] Finished at: 2023-05-05T16:56:56Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jenkins-ci.tools:maven-hpi-plugin:3.44-SNAPSHOT:resolve-test-dependencies (default-cli) on project dependant: Unable to resolve dependencies: Could not resolve dependencies for project org.jenkins-ci.tools.hpi.its:dependant:hpi:1.0-SNAPSHOT: The following artifacts could not be resolved: org.jenkins-ci.tools.hpi.its:override-war-with-snapshot:jar:2.0.0-SNAPSHOT (private-e493f342-jenkins) (absent): Could not find artifact org.jenkins-ci.tools.hpi.its:override-war-with-snapshot:jar:2.0.0-SNAPSHOT (private-e493f342-jenkins) in do-proxy-incrementals (https://repo.do.jenkins.io/incrementals/) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.jenkins-ci.tools:maven-hpi-plugin:3.44-SNAPSHOT:resolve-test-dependencies (default-cli) on project dependant: Unable to resolve dependencies
```

* [passes](https://github.com/jenkinsci/maven-hpi-plugin/pull/480/checks?check_run_id=13269639437) with all fixes. (#481 and #478) 

### Submitter checklist

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/it/override-war-with-snapshot/dependant/src/test/java/org/jenkins_ci/tools/hpi/its/dependant/DependencyTest.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import java.io.IOException;
+import org.jenkins_ci.tools.hpi.its.override_war_with_snapshot.Version;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class DependencyTest {
+
+    @Test
+    public void testVersionOfDependency() throws IOException {
+        Assert.assertEquals("2.0.0-SNAPSHOT", Version.getVersion());
```
Comment (jtnord) [inline src/it/override-war-with-snapshot/dependant/src/test/java/org/jenkins_ci/tools/hpi/its/dependant/DependencyTest.java:12]: test will fail when run using the dependency declared in the pom, but will pass when using the version of the plugin in the war.

File: src/it/override-war-with-snapshot/invoker.properties [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Pass the version of the hpi-plugin here and not in the pom, as pom filtering requires aggregator/child relationship
+
+# First just build the plugin that we depend on
+invoker.goals.1=-f war-with-plugin -ntp -Dhpi-plugin.version=${project.version} clean install -P quick-build -Drevision=1.0.0-SNAPSHOT
+# Then build another version and create a war containing the newer version
+invoker.goals.2=-f war-with-plugin -ntp -Dhpi-plugin.version=${project.version} clean install -P quick-build -P build-war -Drevision=2.0.0-SNAPSHOT
```
Comment (jtnord) [inline src/it/override-war-with-snapshot/invoker.properties:6]: slightly abnormal syntax of profile activation as invoker will split into multiple invocations on seeing a `,` so we use `-P` twice

File: src/it/override-war-with-snapshot/verify.groovy [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import java.io.File
+import java.util.jar.JarFile
+
+assert new File(basedir, 'war-with-plugin/target/override-war-with-snapshot.war').exists()
+
+jf = new JarFile(new File(basedir, 'war-with-plugin/target/override-war-with-snapshot.war'));
+
+assert jf.getEntry('WEB-INF/plugins/override-war-with-snapshot.hpi') != null
+
+return true;
```
Comment (jtnord) [inline src/it/override-war-with-snapshot/verify.groovy:10]: we do not need to perform any checking of the last invocation of the dependant plugin as the junit test will fail if the dependency is not updated.

File: src/it/override-war-with-snapshot/war-with-plugin/pom.xml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            <artifactId>maven-hpi-plugin</artifactId>
+            <extensions>true</extensions>
+            <executions>
+              <execution>
+                <id>build-war-with-plugin</id>
+                <goals>
+                  <goal>custom-war</goal>
+                </goals>
+                <configuration>
+                  <addThisPluginToCustomWar>true</addThisPluginToCustomWar>
```
Comment (jtnord) [inline src/it/override-war-with-snapshot/war-with-plugin/pom.xml:58]: requires #479

File: src/it/override-war-with-snapshot/dependant/pom.xml
Comment (jtnord) [inline src/it/override-war-with-snapshot/dependant/pom.xml:1]: this plugin depends on the plugin in `war-with-plugin`

File: src/it/override-war-with-snapshot/war-with-plugin/src/main/java/org/jenkins_ci/tools/hpi/its/override_war_with_snapshot/Version.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package org.jenkins_ci.tools.hpi.its.override_war_with_snapshot;
+
+import java.io.InputStream;
+import java.io.IOException;
+import java.util.Properties;
+
+public class Version {
+
+    public static String getVersion() throws IOException {
```
Comment (jtnord) [inline src/it/override-war-with-snapshot/war-with-plugin/src/main/java/org/jenkins_ci/tools/hpi/its/override_war_with_snapshot/Version.java:9]: will return a different version depending on the version of the plugin.

File: pom.xml [hunk 6]
```diff
@@ -372,7 +372,7 @@
           <localRepositoryPath>${basedir}/target/local-repo</localRepositoryPath>
           <settingsFile>src/it/settings.xml</settingsFile>
           <pomIncludes>
-            <pomInclude>*/pom.xml</pomInclude>
+            <pomInclude>*</pomInclude>
```
Comment (jtnord) [inline pom.xml:375]: any subdirectory (non recursive) of `src/it`  is now a valid IT - we do not need a `pom.xml` in the root which allows us to have multiple projects for a single IT that do not share a common aggregator

File: src/main/java/org/jenkinsci/maven/plugins/hpi/WarMojo.java
Comment (jtnord) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/WarMojo.java:1]: this file is #481

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (jtnord) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:1]: this file is #478

File: src/it/override-war-with-snapshot/war-with-plugin/pom.xml [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    <license>
+      <name>MIT License</name>
+      <url>https://opensource.org/licenses/MIT</url>
+    </license>
+  </licenses>
+
+  <developers>
+    <developer>
+      <id>Bob</id>
+      <name>Bob McBobfaceNoam Chomsky</name>
```
Comment (basil) [inline src/it/override-war-with-snapshot/war-with-plugin/pom.xml:30]: Multiple names when a single name is expected

```suggestion
      <name>Bob McBobface</name>
```

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java [hunk 8]
```diff
@@ -550,6 +549,12 @@ private static Map<String, String> scanWar(File war, MavenSession session, Maven
                         if (version == null) {
                             throw new IllegalArgumentException("Failed to determine version for " + name);
                         }
+                        // handle any extra info in snapshots e.g. " (private-abcd1234-username)"
+                        version = version.replaceFirst(" [(].+[)]$", "");
+                        if (version.endsWith("-SNAPSHOT")) {
+                            getLog().warn("War contains a SNAPSHOT of " + groupId + ":" + artifactId
+                                    + " build will not be fully repeatable");
```
Comment (basil) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:556]: WAR is capitalized incorrectly, and there is a run-on sentence without punctuation:

```suggestion
                            getLog().warn("WAR contains a SNAPSHOT of " + groupId + ":" + artifactId
                                    + "; build will not be fully repeatable");
```

**Review by basil — APPROVED** [review]

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java (hunk 8 above)
Comment (jtnord) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:556]: addressed in https://github.com/jenkinsci/maven-hpi-plugin/pull/478

**Pull request merged**